### PR TITLE
Add package level documentation for db, historydb, l2db, api, apitype…

### LINKED ISSUE #679: Add package level documentation for api, apitypes, stateapiupdater, historydb and l2db

### PR DESCRIPTION
**Pull request by arnaubennassar**
Close #679

### CODE REVIEW

File: db/historydb/historydb.go [hunk 1]
```diff
@@ -1,5 +1,27 @@
...
+Some of the database tooling used in this package such as meddler and migration tools is explained at the db package.
+
+This package is spitted in different files following this ideas:
+- historydb.go: constructor and functions used by packages other than the api.
+- apiqueries.go: functions used by the API, the queries implemented in this functions use a semaphore
+to restrict the maximum concurrent connections to the database.
+- views.go: structs used to retrieve/store data from/to the database. When possible, the common structs are used, however
+most of the times there is no 1:1 relation between the struct fields and the tables of the schema, specially when joining tables.
+In some cases some of the structs defined in this file also include custom Marshallers to easily match the expected api formats.
+- nodeinfo.go: used to handle the interfaces and structs that allow communication across running in different machines/process but sharing the same database.
```
Comment (Pantani) [inline db/historydb/historydb.go:22]:
```suggestion
Package historydb is responsible for storing and retrieving the historic data of the Hermez network.
It's mostly but not exclusively used by the API and the synchronizer.

Apart from the logic defined in this package, it's important to notice that there are some triggers defined in the
migration files that have to be taken into consideration to understanding the results of some queries. This especially true
for reorgs: all the data is directly or indirectly related to a block, this makes handling reorgs as easy as deleting the
reorged blocks from the block table, and all related items will be dropped in cascade. This is not the only case, in general
functions defined in this package that get affected somehow by the SQL level defined logic has a special mention on the function description.

Some of the database tooling used in this package such as meddler and migration tools is explained in the db package.

This package is spitted in different files following these ideas:
- historydb.go: constructor and functions used by packages other than the api.
- apiqueries.go: functions used by the API, the queries implemented in this functions use a semaphore
to restrict the maximum concurrent connections to the database.
- views.go: structs used to retrieve/store data from/to the database. When possible, the common structs are used, however
most of the time there is no 1:1 relation between the struct fields and the tables of the schema, especially when joining tables.
In some cases, some of the structs defined in this file also include custom Marshallers to easily match the expected api formats.
- nodeinfo.go: used to handle the interfaces and structs that allow communication across running in different machines/process but sharing the same database.
```

File: db/utils.go [hunk 2]
```diff
@@ -1,5 +1,13 @@
 package db
 
+/*
+Package db have some common utilities shared by db/l2db and db/historydb, most relevant ones are:
+- SQL connection utilities
+- Mannaging the SQL schema: this is done using migration files placed under db/migrations. The files are executed by
+order of the file name.
+- Custom meddlers: used to easily transform struct <==> table
```
Comment (Pantani) [inline db/utils.go:8]:
```suggestion
Package db have some common utilities shared by db/l2db and db/historydb, the most relevant ones are:
- SQL connection utilities
- Managing the SQL schema: this is done using migration files placed under db/migrations. The files are executed by
order of the file name.
- Custom meddlers: used to easily transform struct <==> table
```

File: db/l2db/l2db.go [hunk 3]
```diff
@@ -1,5 +1,23 @@
...
+
+Some of the database tooling used in this package such as meddler and migration tools is explained at the db package.
+
+This package is spitted in different files following this ideas:
+- l2db.go: constructor and functions used by packages other than the api.
+- apiqueries.go: functions used by the API, the queries implemented in this functions use a semaphore
+to restrict the maximum concurrent connections to the database.
+- views.go: structs used to retrieve/store data from/to the database. When possible, the common structs are used, however
+most of the times there is no 1:1 relation between the struct fields and the tables of the schema, specially when joining tables.
+In some cases some of the structs defined in this file also include custom Marshallers to easily match the expected api formats.
```
Comment (Pantani) [inline db/l2db/l2db.go:18]:
```suggestion
Package l2db is responsible for storing and retrieving the data received by the coordinator through the api.
Note that this data will be different for each coordinator in the network, as this represents the L2 information.

The data managed by this package is fundamentally PoolL2Tx and AccountCreationAuth. All this data come from
the API sent by clients and is used by the txselector to decide which transactions are selected to forge a batch.

Some of the database tooling used in this package such as meddler and migration tools is explained in the db package.

This package is spitted in different files following these ideas:
- l2db.go: constructor and functions used by packages other than the api.
- apiqueries.go: functions used by the API, the queries implemented in this functions use a semaphore
to restrict the maximum concurrent connections to the database.
- views.go: structs used to retrieve/store data from/to the database. When possible, the common structs are used, however
most of the time there is no 1:1 relation between the struct fields and the tables of the schema, especially when joining tables.
In some cases, some of the structs defined in this file also include custom Marshallers to easily match the expected api formats.
```

File: api/stateapiupdater/stateapiupdater.go [hunk 4]
```diff
@@ -1,5 +1,13 @@
 package stateapiupdater
 
+/*
+Package stateapiupdater is responsible for generating and storing the object response of the GET /state endpoint exposed through the api package.
+This object is extensively defined at the OpenAPI spec located at api/swagger.yml.
+
+Deployment considerations: in a setup were multiple process are used (dedicated api process, separated coord / sync, ...) only one process should care
+of using this package.
```
Comment (Pantani) [inline api/stateapiupdater/stateapiupdater.go:8]:
```suggestion
Package stateapiupdater is responsible for generating and storing the object response of the GET /state endpoint exposed through the api package.
This object is extensively defined at the OpenAPI spec located at api/swagger.yml.

Deployment considerations: in a setup where multiple processes are used (dedicated api process, separated coord / sync, ...), only one process should care
of using this package.
```

File: api/api.go [hunk 5]
```diff
@@ -1,5 +1,34 @@
...
+the system is done through the SQL database. As a matter of fact, the only public function of the package is the constructor NewAPI.
+All the information needed for the API to work should be obtained throgh the configuration file of the cli or the database.
+- The format of the requests / responses doesn't match directly with the common types and for this reason the apckage api/apitypes is used
+to facilitate the format conversion. Most of the times this is done directly at the db level.
+- The API endpoints are fully documented using OpenAPI fka Swagger, all the endpoints are tested against the spec to ensure consistency
+between implementation and specification. To get a sense of which endpoints exist and how they work it's strongly recommended to check this specification.
+The specification can be found at api/swagger.yml.
+- In general all the API endpoints produce queries to the SQL database in order to retrieve / insert the requested information. The most notable exceptions to this are
+the /config endpoint which return a static object generated at construction time, and the /state which also is retrieved from the database, but it's generated by api/stateapiupdater package.
```
Comment (Pantani) [inline api/api.go:29]:
```suggestion
Package api implements the public interface of the hermez-node using a HTTP REST API.

There are two subsets of endpoints:

- coordinatorEndpoints: used to receive L2 transactions and account creation authorizations. Targeted for wallets.
- explorerEndpoints: used to provide all sorts of information about the network. Targeted for explorers and similar services.

About the configuration of the API:

- The API is supposed to be launched using the cli found at the package cli/node, and configured through the configuration file.
- The mentioned configuration file allows exposing any combination of the endpoint subsets.
- Although the API can run in a "standalone" manner using the serveapi command, it won't work properly
unless another process acting as a coord or sync is filling the HistoryDB.

Design principles and considerations:

- In order to decouple the API process from the rest of the node, all the communication between this package and the rest of
the system is done through the SQL database. As a matter of fact, the only public function of the package is the constructor NewAPI.
All the information needed for the API to work should be obtained through the configuration file of the cli or the database.
- The format of the requests / responses doesn't match directly with the common types, and for this reason, the package api/apitypes is used
to facilitate the format conversion. Most of the time, this is done directly at the db level.
- The API endpoints are fully documented using OpenAPI fka Swagger. All the endpoints are tested against the spec to ensure consistency
between implementation and specification. To get a sense of which endpoints exist and how they work, it's strongly recommended to check this specification.
The specification can be found at api/swagger.yml.
- In general, all the API endpoints produce queries to the SQL database in order to retrieve / insert the requested information. The most notable exceptions to this are
the /config endpoint, which returns a static object generated at construction time, and the /state, which also is retrieved from the database, but it's generated by API/stateapiupdater package.
```

**Review by Pantani — CHANGES_REQUESTED** [review]
This is a good tool to help us to avoid this typo stuffs:

https://app.grammarly.com/

**Review by ed255 — APPROVED** [review]

File: api/api.go [hunk 6]
```diff
@@ -1,5 +1,30 @@
 package api
 
+/*
```
[the remark this answers is not in the document]
Comment (ed255) [reply api/api.go:3]: The same applies to the documentation you added to the other packages!

**Review by ed255 — CHANGES_REQUESTED** [review]

File: api/api.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (arnaubennassar) [reply api/api.go:3]: That was my intention, but after saving the file my formater would add this empty line :S

File: api/api.go [hunk 7]
```diff
@@ -1,5 +1,30 @@
...
+- Although the API can run in a "standalone" manner using the serveapi command, it won't work properly
+unless another process acting as a coord or sync is filling the HistoryDB.
+
+Design principles and considerations:
+- In order to decouple the API process from the rest of the node, all the communication between this package and the rest of
+the system is done through the SQL database. As a matter of fact, the only public function of the package is the constructor NewAPI.
+All the information needed for the API to work should be obtained through the configuration file of the cli or the database.
+- The format of the requests / responses doesn't match directly with the common types, and for this reason, the package api/apitypes is used
+to facilitate the format conversion. Most of the time, this is done directly at the db level.
+- The API endpoints are fully documented using OpenAPI fka Swagger. All the endpoints are tested against the spec to ensure consistency
```
[the remark this answers is not in the document]
Comment (arnaubennassar) [reply api/api.go:21]: `fka` stands for Formerly Known As, which is more accurate than, although if this acronym is not that well known maybe it's better to go with `aka`

**Review by Pantani — APPROVED** [review]

**Review by ed255 — APPROVED** [review]

**Review by ed255 — APPROVED** [review]

**Review by arnaucube — APPROVED** [review]

**Pull request merged**